Modules/posixmodule.c

```diff
@@ -10069,9 +10069,20 @@ os__getdiskusage_impl(PyObject *module, path_t *path)
 {
     BOOL retval;
     ULARGE_INTEGER _, total, free;
+    WCHAR wpath[MAX_PATH];
+    struct _Py_stat_struct stat;
+
+    assert(path->length + 1 < MAX_PATH);
+    wcscpy_s(wpath, path->length + 1, path->wide);
+
+    if (win32_stat(wpath, &stat))
```

Comment:
`win32_stat` requires multiple system calls, so it's relatively expensive. I think only the uncommon case should be forced to pay this penalty, i.e. retry on the parent directory if `GetDiskFreeSpaceExW` fails with `GetLastError() == ERROR_DIRECTORY`. If you prefer to check first, call `GetFileAttributesW` instead to check for `FILE_ATTRIBUTE_DIRECTORY`, which is a relatively cheap call.

Reply:
Wow - thanks for taking a look so quickly, and thanks for the advice! I updated the diff to use a `GetFileAttributesW` check first, since it keeps the error checking logic simple. If you'd still rather I only do the fix if `GetDisFreeSpaceExW` fails, I'll be happy to update the diff again to do so.